kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/RecordEncryptor.java

```diff
@@ -0,0 +1,148 @@
+    private ByteBuffer transformedValue;
+    private Header[] transformedHeaders;
+
+    /**
+     * Constructor (obviously).
+     * @param encryptionVersion The encryption version
+     * @param encryptionScheme The encryption scheme for this key
+     * @param keyContext The key context
+     * @param parcelBuffer A buffer big enough to write the parcel
+     * @param wrapperBuffer A buffer big enough to write the wrapper
+     */
+    RecordEncryptor(EncryptionVersion encryptionVersion,
+                    EncryptionScheme<K> encryptionScheme,
+                    KeyContext keyContext,
+                    ByteBuffer parcelBuffer,
+                    ByteBuffer wrapperBuffer) {
+        this.encryptionVersion = encryptionVersion;
+        this.encryptionScheme = encryptionScheme;
+        this.keyContext = keyContext;
+        this.parcelBuffer = parcelBuffer;
+        this.wrapperBuffer = wrapperBuffer;
+        this.encryptionHeader = new Header[]{ new RecordHeader(ENCRYPTION_HEADER_NAME, new byte[]{ encryptionVersion.code() }) };
+    }
+
+    @Override
+    public void init(@NonNull Record kafkaRecord) {
+        if (encryptionScheme.recordFields().contains(RecordField.RECORD_HEADER_VALUES)
+                && kafkaRecord.headers().length > 0
+                && !kafkaRecord.hasValue()) {
+            // todo implement header encryption preserving null record-values
+            throw new IllegalStateException("encrypting headers prohibited when original record value null, we must preserve the null for tombstoning");
+        }
+        if (kafkaRecord.hasValue()) {
+            Parcel.writeParcel(encryptionVersion.parcelVersion(), encryptionScheme.recordFields(), kafkaRecord, parcelBuffer);
+            parcelBuffer.flip();
+            this.transformedValue = writeWrapper(parcelBuffer);
+            parcelBuffer.rewind();
+        }
+        else {
+            this.transformedValue = null;
+        }
+
+        Header[] oldHeaders = kafkaRecord.headers();
+        if (encryptionScheme.recordFields().contains(RecordField.RECORD_HEADER_VALUES) || oldHeaders.length == 0) {
+            transformedHeaders = encryptionHeader;
+        }
+        else {
+            transformedHeaders = new Header[1 + oldHeaders.length];
+            transformedHeaders[0] = encryptionHeader[0];
+            System.arraycopy(oldHeaders, 0, transformedHeaders, 1, oldHeaders.length);
+        }
+    }
+
+    @Nullable
+    private ByteBuffer writeWrapper(ByteBuffer parcelBuffer) {
+        switch (encryptionVersion.wrapperVersion()) {
+            case V1 -> {
+                var edek = keyContext.serializedEdek();
+                ByteUtils.writeUnsignedVarint(edek.length, wrapperBuffer);
+                wrapperBuffer.put(edek);
+                wrapperBuffer.put(AadSpec.NONE.code()); // aadCode
+                wrapperBuffer.put(CipherCode.AES_GCM_96_128.code());
+                keyContext.encodedSize(parcelBuffer.limit());
```

Comment:
think this might be dead code from when we were writing the encoded size into the wrapper?

Reply:
IIRC the KeyContext is stateful in this regard, and if you ask it to encrypt without having asked it for the size the its internal state is wrong and you'll get an exception.